### PR TITLE
Harden environment startup

### PR DESCRIPTION
**Pull request by wendigo**
Sometimes environments fail to start because containers startup depend on each other

### CODE REVIEW

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 1]
```diff
@@ -131,6 +131,11 @@ private Environment tryStart()
             }
 
             Startables.deepStart(containers).get();
+            if (!allContainersHealthy(containers)) {
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:134]: Add a comment why we want to do extra check here

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 2]
```diff
@@ -322,6 +322,15 @@ public String getEnvironmentName()
             return name;
         }
 
+        public Builder containerDependsOn(String container, String otherContainer)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:325]: parameter naming does not make it obvious which container depends on which.
Maybe name `container` and `dependencyContainer`?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 3]
```diff
@@ -322,6 +322,15 @@ public String getEnvironmentName()
...
+        {
+            checkState(containers.containsKey(container), "Container with name %s is not registered", name);
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:327]: Also instead we could maybe add `getContainer` method to builder and go with:
```java
        builder.addContainers(createZookeeper(), createKafka())
                .configureContainer("kafka", kafkaContainer -> kafkaContainer.dependsOn(builder.getContainer("zookeeper")));
```

But I think helper method is nice

**Review by losipiuk — APPROVED** [review]

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java (hunk 3 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:327]: That was my initial thought. I dont't know if builder should have get method though (what's your take on that?)

Comment (losipiuk) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:327]: I would not object if there are more cases when that would be useful.
But the helper method you added is fine too. And it does `checkState` calls which would make the `configureContainer` bloated.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 4]
```diff
@@ -284,6 +286,14 @@ public void copyLogsToHostPath(Path hostPath)
         return statistics.get();
     }
 
+    public DockerContainer waitingForAll(WaitStrategy... strategies)
+    {
+        WaitAllStrategy strategy = new WaitAllStrategy();
+        Arrays.stream(strategies).forEach(strategy::withStrategy);
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:292]: make it simple for loop

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/common/Hadoop.java [hunk 5]
```diff
@@ -98,7 +99,7 @@ public static DockerContainer createHadoopContainer(DockerFiles dockerFiles, Str
                 .withCommand("/usr/local/hadoop-run.sh")
                 .withExposedLogPaths("/var/log/hadoop-yarn", "/var/log/hadoop-hdfs", "/var/log/hive", "/var/log/container-health.log")
                 .withStartupCheckStrategy(new IsRunningStartupCheckStrategy())
-                .waitingFor(new SelectedPortWaitStrategy(10000)) // HiveServer2
+                .waitingForAll(forSelectedPorts(10000), forHealthcheck()) // HiveServer2
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/common/Hadoop.java:102]: you are adding `forHealthCheck()` here. SHould be separate commit.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/common/Standard.java [hunk 6]
```diff
@@ -125,7 +126,7 @@ public static DockerContainer createPrestoContainer(DockerFiles dockerFiles, Fil
                 .withFileSystemBind(serverPackage.getPath(), "/docker/presto-server.tar.gz", READ_ONLY)
                 .withCommand("/docker/presto-product-tests/run-presto.sh")
                 .withStartupCheckStrategy(new IsRunningStartupCheckStrategy())
-                .waitingFor(Wait.forLogMessage(".*======== SERVER STARTED ========.*", 1))
+                .waitingForAll(forLogMessage(".*======== SERVER STARTED ========.*", 1), forHealthcheck())
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/common/Standard.java:129]: same here

**Review by losipiuk — APPROVED** [review]

**Review by losipiuk — APPROVED** [review]
